Reduce permission of antrea-agent service account

xliuxu opened the pull request:
Remove the update permission for services/status of antrea-agent
service account. Remove the optimization for ExternalTrafficPolicy
setting to Local cases in ServiceExternalIP feature accordingly.

Signed-off-by: Xu Liu <xliu2@vmware.com>

tnqn reviewed (dismissed):
LGTM

xliuxu commented:
@tnqn The e2e test failed because with this change the testing code cannot determine which Node owns the external IP. Most of the test cases will need to check the assigned Node by the hostname field in the status of a Service. I cannot find an easy way to figure out which Node owns the IP as they are not physically assigned. Do you think it is OK to remove all related test cases until we have a new API for assigned Node checking?

antoninbas commented:
@xliuxu or maybe you meant "unit tests failed" instead of "e2e tests failed"?

xliuxu commented:
@antoninbas No. For LB IP of Services, we do not assign IP to any interfaces on the Node. Otherwise, it will conflict with the implementation of proxyAll feature. We use userspace ARP/NDP responders to handle ARP/NDP queries instead.

tnqn commented:
@xliuxu we need a way to check which Node owns the IP anyway for troubleshooting anyway. Could we add a `antctl get serviceexternalip` command to agent mode to dump something like the following?
```
NAMESPACE  NAME        EXTERNAL-IP   NODE
default    service-1   1.1.1.1       worker-1
default    service-2   1.1.1.2       worker-2
```

xliuxu commented:
@tnqn  Thanks for the suggestion. I have added a new command `antctl get serviceexternalip` for checking the status of the assigned Node and adjusted the e2e codes.

tnqn commented on pkg/querier/querier.go:
```diff
@@ -90,3 +90,17 @@ type NetworkPolicyQueryFilter struct {
+	GetServiceExternalIPStatus() []ServiceExternalIPInfo
+}
+
+// ServiceExternalIPInfo is used to check the status of external IPs for Services.
```
This sounds like the comment of `ServiceExternalIPStatusQuerier`

xliuxu replied:
updated.

tnqn reviewed (commented):
LGTM overall

xliuxu commented on cmd/antrea-agent/agent.go:
```diff
@@ -597,6 +597,7 @@ func run(o *Options) error {
 		ovsBridgeClient,
 		proxier,
 		networkPolicyController,
+		externalIPController,
```
done.

tnqn reviewed (dismissed):
LGTM

tnqn commented:
/test-all

tnqn commented:
@antoninbas will you take another look?

antoninbas commented on pkg/agent/controller/serviceexternalip/controller.go:
```diff
@@ -337,68 +339,51 @@ func (c *ServiceExternalIPController) syncService(key apimachinerytypes.Namespac
+	nodeName, err := c.cluster.SelectNodeForIP(currentExternalIP, ipPool, filters...)
+	if err != nil {
+		if err == memberlist.ErrNoNodeAvailable {
+			// No Node is available for the moment. The Service will be requeued by Endpoints, Node, or Memberlist update events.
```
s/for the moment/at the moment

antoninbas commented on pkg/antctl/antctl.go:
```diff
@@ -509,6 +510,32 @@ var CommandList = &commandList{
+					params: []flagInfo{
+						{
+							name:  "name",
+							usage: "Retrieve Service by name. If present, Namespace must be provided.",
```
```suggestion
							usage: "Name of the Service; if present, Namespace must be provided as well.",
```

antoninbas commented on pkg/antctl/antctl.go:
```diff
@@ -509,6 +510,32 @@ var CommandList = &commandList{
+						},
+						{
+							name:      "namespace",
+							usage:     "Get Services from specific Namespace.",
```
```suggestion
							usage:     "Only get the external IP status for Services in the provided  Namespace.",
```

antoninbas commented on pkg/agent/apiserver/apiserver.go:
```diff
@@ -82,6 +83,7 @@ func installHandlers(aq agentquerier.AgentQuerier, npq querier.AgentNetworkPolic
 	s.Handler.NonGoRestfulMux.HandleFunc("/addressgroups", addressgroup.HandleFunc(npq))
 	s.Handler.NonGoRestfulMux.HandleFunc("/ovsflows", ovsflows.HandleFunc(aq))
 	s.Handler.NonGoRestfulMux.HandleFunc("/ovstracing", ovstracing.HandleFunc(aq))
+	s.Handler.NonGoRestfulMux.HandleFunc("/serviceexternalip", serviceexternalip.HandleFunc(seipq))
```
you may want to add the `/serviceexternalip` URL to the `antctl` RBAC ClusterRole, just for consistency (and in case someone is using the `antctl` ServiceAccount token to access antrea APIs directly)

xliuxu replied:
added.

antoninbas commented on pkg/agent/controller/serviceexternalip/controller.go:
```diff
@@ -453,3 +438,19 @@ func (c *ServiceExternalIPController) nodesHasHealthyServiceEndpoint(service *co
+func (c *ServiceExternalIPController) GetServiceExternalIPStatus() []querier.ServiceExternalIPInfo {
+	c.externalIPStatesMutex.RLock()
+	defer c.externalIPStatesMutex.RUnlock()
+	info := make([]querier.ServiceExternalIPInfo, len(c.externalIPStates))
+	idx := 0
+	for k, v := range c.externalIPStates {
+		info[idx].ServiceName = k.Name
+		info[idx].Namespace = k.Namespace
+		info[idx].ExternalIP = v.ip
+		info[idx].ExternalIPPool = v.ipPool
+		info[idx].AssignedNode = v.assignedNode
+		idx++
+	}
```
nit: I think something like this is more elegant:
```go
info := make([]querier.ServiceExternalIPInfo, 0, len(c.externalIPStates))
for k, v := range c.externalIPStates {
    info = append(info, querier.ServiceExternalIPInfo{
        ServiceName: k.Name,
        // ...
    })
```

xliuxu replied:
thanks. updated.

antoninbas commented on pkg/agent/controller/serviceexternalip/controller.go:
```diff
@@ -71,7 +71,7 @@ type ServiceExternalIPController struct {
 	queue workqueue.RateLimitingInterface
 
-	externalIPStates      map[apimachinerytypes.NamespacedName]externalIPState
+	externalIPStates      map[apimachinerytypes.NamespacedName]*externalIPState
```
is there a reason for making this is a pointer? Based on usage in `GetServiceExternalIPStatus`, it doesn't seem necessary.

xliuxu replied:
It is just for the defer statement to save the `externalIPState`. I have reverted this change since it is not necessary and can also be achieved by changing the signature of `func saveServiceState`.

antoninbas reviewed (commented):
some small comments and questions, but overall LGTM

tnqn reviewed (approved):
LGTM

xliuxu commented on test/e2e/antctl_test.go:
```diff
@@ -113,7 +113,7 @@ func testAntctlAgentLocalAccess(t *testing.T, data *TestData) {
 		t.Run(cmd, func(t *testing.T) {
 			stdout, stderr, err := runAntctl(podName, args, data)
-			if err != nil {
+			if err != nil && !strings.HasSuffix(stderr, "not enabled\n") {
```
@tnqn  I just pushed a new commit with a single line change here to fix the e2e error for `antctl`.

tnqn commented:
/test-all
/test-ipv6-all
/test-ipv6-only-all
/test-windows-all

xliuxu commented:
/test-networkpolicy 
/test-windows-conformance 
/test-ipv6-only-networkpolicy
/test-ipv6-e2e

antoninbas commented:
@xliuxu please cherry-pick this to release-1.6
